### ISSUE TITLE
Cannot execute consecutive queries that would return the same `data.

### ISSUE THREAD

**Issue by nfantone** [convoyinc/apollo-cache-hermes#405]
> This is a follow up to https://github.com/apollographql/react-apollo/issues/1931

**Summary**:
If executing the same query two (or more) consecutive times **that would return the same `data`**, observers listening to events from `apollo-client` `client.watchQuery` are not notified of the change. This leads to necessary UI updates not occurring. In particular, `<Query>` component from `react-apollo` does not re-render and stays in "loading" state.

Here is a repro case:

[![Edit apollo-client-error-template](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/lpqzpkoop7)

1. Click on "Run query".
2. Query is executed, people list is updated.
3. Click on "Run query" again to trigger a second query that would fetch the same data.
4. Notice how UI gets stuck in "Loading...".

Now, go to `index.js`, switch `cache` to `apollo-inmemory-cache` and try above again. It'll work just fine.

**Expected Behavior**:

- Observers, such as `react-apollo`'s `Query` component, are _always_ notified of the the query's network result, regardless of its content.
- `apollo-inmemory-cache` functionality is matched.

**Actual Behavior**:

Cannot run two (or more) consecutive queries that would return the exact same `data.

**Comment by danieldunderfelt** [comment]
I'm seeing the same issue. I initially thought it only happens when the server returns empty data, but good to know that it also happens for any data that is the same as last time. WIll switch to inmemory for now.

**Comment by nfantone** [comment]
Sad that we are not hearing anything from owners. This is a deal-breaker for us on switching to `apollo-cache-hermes`.

**Comment by wmertens** [comment]
I wonder what causes this. There must be some side-effect that apollo-client is counting on to update the loading state?

**Comment by dios-david** [comment]
After taking a quick look into this issue, I found that the ObservableQuery implementation in apollo-client has a method called [isDifferentFromLastResult](https://github.com/apollographql/apollo-client/blob/master/packages/apollo-client/src/core/ObservableQuery.ts#L255) which compares the previous and the new results when a query change happens. If the `networkStatus` + `stale` + `data` properties are equal in both results, the update won't be triggered.

When I change a query using *apollo-cache-inmemory* this method is being called twice:
*initial state:* `loading: false, networkStatus: 7, data: <oldData>`
(fetching data...)
*1st call:* `loading: true, networkStatus: 2, data: <oldData>`
(fetching data finished)
*2nd call:* `loading: false, networkStatus: 7, data: <newData>`

In my case the data is the same (oldData = newData), but this method will return `true` in both cases, because the `loading` and the `networkStatus` properties have different values.

When I change a query using *apollo-cache-hermes* this method is being called only once:
*initial state:* `loading: false, networkStatus: 7, data: <oldData>`
(fetching data...)
(fetching data finished)
*1st call:* `loading: false, networkStatus: 7, data: <newData>`

Because oldData=newData, and `loading` and `networkStatus` has the same values as the previous result, this method will return `true` so no update will be triggered.

Basically, the "loading state" result is missing when using apollo-cache-hermes, that's why this logic in apollo-client breaks.
